ktor-network/jvm/src/io/ktor/network/sockets/DatagramSendChannel.kt

```diff
@@ -48,41 +50,81 @@ internal class DatagramSendChannel(
         return true
     }
 
-    @OptIn(InternalCoroutinesApi::class)
+    @OptIn(InternalCoroutinesApi::class, InternalIoApi::class, UnsafeIoApi::class)
     override fun trySend(element: Datagram): ChannelResult<Unit> {
         if (!lock.tryLock()) return ChannelResult.failure()
 
-        var result = false
-
         try {
-            DefaultDatagramByteBufferPool.useInstance { buffer ->
-                element.packet.copy().readAvailable(buffer)
-                result = channel.send(buffer, element.address.toJavaAddress()) == 0
+            val packetSize = element.packet.remaining
+            var writeWithPool = false
+            UnsafeBufferOperations.readFromHead(element.packet.buffer) { buffer ->
+                val length = buffer.remaining()
+                if (length < packetSize) {
+                    // Packet is too large to read directly.
+                    writeWithPool = true
+                    return@readFromHead
+                }
+
+                val result = channel.send(buffer, element.address.toJavaAddress()) == 0
+                if (result) {
+                    buffer.position(buffer.limit())
+                } else {
+                    buffer.position(0)
+                }
+            }
+            if (writeWithPool) {
+                DefaultDatagramByteBufferPool.useInstance { buffer ->
+                    element.packet.peek().writeMessageTo(buffer)
+
+                    val result = channel.send(buffer, element.address.toJavaAddress()) == 0
+                    if (result) {
+                        element.packet.discard()
+                    }
+                }
             }
         } finally {
             lock.unlock()
         }
 
-        if (result) {
-            element.packet.close()
```

Comment:
close actually does not recycle the head. Now fixed with `buffer.position(buffer.limit())` or `element.packet.discard()`.